### PR TITLE
Update LLVM to c68b560be381e831fd72c6b8b8909427e4e2ff36.

### PR DESCRIPTION
**Pull request by mikeurbach**
Some changes in this bump:
* `OpBuilderDag` -> `OpBuilder`: https://github.com/llvm/llvm-project/commit/32c49c7d730d8e46921f3f17aee52a7a5320d9ef
* Changes to `Type` `getChecked` and verification: https://github.com/llvm/llvm-project/commit/06e25d56451977ef5b7052282faacfe3d42acb65
* Changes to `Value` internals, with minor public breakage: https://github.com/llvm/llvm-project/commit/3dfa86149e14ebb7fa40c55749ae435641940ff6
* Changes to LLVM translation: https://github.com/llvm/llvm-project/commit/19db802e7b13f190ebae6f5389b0bfbc1827e4e7

### CODE REVIEW

**Comment by mikeurbach** [comment]
Seemed mostly mechanical, and the test passed locally on this one. NPComp had some issues though, so I'm running the PR checks as well as the integration checks.

**Review by teqdruid — APPROVED** [review]
Agreed this is mechanical. LGTM!

**Comment by mikeurbach** [comment]
Looks like Clang release build passed, but GCC debug didn't due to some warnings that are really errors.. Will update.

**Comment by mikeurbach** [comment]
Strangely enough the PR gate tests pass, but I can't seem to get the same checks to pass in the integration tests: https://github.com/llvm/circt/runs/2083167055. Something about `MLIR Parser: custom op parser 'llhd.entity'`. Trying to repro that failure using a local copy of the integration test image...

**Comment by mikeurbach** [comment]
Well, the above change seemed like a longshot, but also seemed to fix the integration tests locally. I still see it here though: https://github.com/llvm/circt/runs/2083773533.

**Comment by mikeurbach** [comment]
The integration tests have passed twice with the last change. Not sure why it was just that file that bothered them. https://github.com/llvm/circt/actions/runs/642058937.

**Comment by mikeurbach** [comment]
PR checks and integration checks are passing, so this looks good.

@youngar it seems like I needed to do this change of splitting success from error tests for one LLHD IR test https://github.com/llvm/circt/commit/bae77c38579353d8df2536fccb63f2f8cdfd585f. I ended up splitting out all the test/Dialect/LLHD/IR tests in this way for good measure, but there may be others in the codebase.

Couple questions: do you have some more background on why this was needed? And specifically, it looks like you removed split-input-file from the success tests, is that something I need to do here?

Since it bit me for that one LLHD IR tests, I kind of want to just resolve this for all the other tests that are like that and be done with it.

**Comment by youngar** [comment]
I'm not totally sure that I made the best change before when I split out the error tests, the real issue was that the output was not being piped to FileCheck. I thought that the output of -verify-diagnostics was somehow not suitable to pipe to FileCheck, but I just checked locally and it looks like everything is fine. I can play around with this some more and see what happens.  Its possible that I was using a newer llvm sha with circt back then, so I'll check the git log for anything fishy.

The primary reason I removed split-input-file was just because it wasn't needed, and I think the test case will have easier to read output when its not split. Not something that you have to do.

**Comment by mikeurbach** [comment]
Ok, thanks for the info.

> I thought that the output of -verify-diagnostics was somehow not suitable to pipe to FileCheck, but I just checked locally and it looks like everything is fine.

I thought we should be able to pipe the output of -verify-diagnostics to FileCheck as well, this works for me locally (I wrote the tests we've had to split initially, and I thought they were working). However, integrations tests were failing last night with some issues related to parsing, and separating the -verify-diagnostics tests from the FileCheck tests seems to resolve this.

I guess at this point I'm thinking it doesn't hurt to just separate these fully where they are currently mixed. I'd keep split-input-file in this case, for the tests that have it.

**Comment by mikeurbach** [comment]
I went ahead and split the rest of the tests that have diagnostics verification mixed with FileCheck tests, since it seemed to fix an issue in the integration tests last night, and I don't want anyone to have to context switch back to this and do the same thing later. If splitting them wasn't really fixing the root of the issue I saw last night... so be it, we can solve that if it happens again.

There are a few tests that intentionally mix diagnostics verification and FileCheck (in the same split, if they're splitting), and I didn't bother pulling them apart. If these cause an issues later, we can deal with it then.

```
test/Conversion/FIRRTLToRTL/lower-to-rtl-module-errors.mlir                                                                                                                                         
test/Dialect/FIRRTL/lower-types-errors.mlir                                                                                                                                                         
test/Dialect/FIRRTL/parse-locations.fir 
```

With the latest, both the PR check and the integration tests pass (https://github.com/llvm/circt/actions/runs/643604335), so I'm going to call this good an land it.

File: lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp [hunk 1]
```diff
@@ -45,10 +46,10 @@ llvm::hash_code computeHash(MemOp op) {
     hash = llvm::hash_combine(hash, resultTypes.front());
     break;
   default:
-    // Use the type buffer as the hash, as we can guarantee it is the same for
+    // Use the type range as the hash, as we can guarantee it is the same for
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp:49]: This change seemed like the correct thing (a `TypeRange` is hashable and comparable, and I think the upstream change did `ArrayRef<Type>` -> `TypeRange` in a similar way). @youngar I think you own this. Do you mind checking here? The changes in this file were the only thing I felt might not be totally mechanical. Is this covered by the existing tests?

Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp:49]: Sorry for the slow response, I missed this question.  You can replace the whole case with the hash of result types/

```diff
+ hash = llvm::hash_combine(hash, op->getResultTypes());
-  // Result Types
-  ArrayRef<Type> resultTypes = op->getResultTypes();
-  switch (resultTypes.size()) {
-  case 0:
-    // We don't need to add anything to the hash.
-    break;
-  case 1:
-    // Add in the result type.
-    hash = llvm::hash_combine(hash, resultTypes.front());
-    break;
-  default:
-    // Use the type buffer as the hash, as we can guarantee it is the same for
-    // any given range of result types. This takes advantage of the fact the
-    // result types >1 are stored in a TupleType and uniqued.
-    hash = llvm::hash_combine(hash, resultTypes.data());
-    break;
-  }
-  return hash
```

File: lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp [hunk 2]
```diff
@@ -85,7 +86,7 @@ static bool isEquivalentTo(MemOp lhs, MemOp rhs) {
     // Use the type buffer for the comparison, as we can guarantee it is the
     // same for any given range of result types. This takes advantage of the
     // fact the result types >1 are stored in a TupleType and uniqued.
-    if (lhsResultTypes.data() != rhsResultTypes.data())
+    if (lhsResultTypes != rhsResultTypes)
       return false;
     break;
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp:91]: Can do the same thing here, replace the whole thing with something similar

```diff
  // Compare result types.
+  if (lhs->getResultTypes() != rhs->getResultTypes())
+    return false;
-  ArrayRef<Type> lhsResultTypes = lhs->getResultTypes();
-  ArrayRef<Type> rhsResultTypes = rhs->getResultTypes();
-  TypeRange lhsResultTypes = lhs->getResultTypes();
-  TypeRange rhsResultTypes = rhs->getResultTypes();
-  if (lhsResultTypes.size() != rhsResultTypes.size())
-    return false;
-  switch (lhsResultTypes.size()) {
-  case 0:
-    break;
-  case 1:
-    // Compare the single result type.
-    if (lhsResultTypes.front() != rhsResultTypes.front())
-      return false;
-    break;
-  default:
-    // Use the type buffer for the comparison, as we can guarantee it is the
-    // same for any given range of result types. This takes advantage of the
-    // fact the result types >1 are stored in a TupleType and uniqued.
-    if (lhsResultTypes.data() != rhsResultTypes.data())
-    if (lhsResultTypes != rhsResultTypes)
-      return false;
-    break;
```

**Comment by mikeurbach** [comment]
Thanks @youngar . Now that I actually look at those snippets in context, totally makes sense.

**Pull request merged**